Composer/packages/client/src/utils/authClient.ts

```diff
@@ -3,34 +3,111 @@
 
 import { AuthParameters } from '@botframework-composer/types';
 
+import { authConfig } from '../constants';
+
+import {
+  getTokenFromCache,
+  createPopupWindow,
+  monitorWindowForQueryParam,
+  createHidenIframe,
+  getIdTokenUrl,
+  getAccessTokenUrl,
+  isTokenExpired,
+  cleanTokenFromCache,
+} from './auth';
+import { isElectron } from './electronUtil';
+import storage from './storage';
+
+let idToken = getTokenFromCache('idToken');
+
 async function getAccessToken(options: AuthParameters): Promise<string> {
+  const { targetResource = '', scopes = [] } = options;
   try {
-    const { clientId = '', targetResource = '', scopes = [] } = options;
-    const { __csrf__ = '' } = window;
+    if (isElectron()) {
+      const { __csrf__ = '' } = window;
 
-    let url = '/api/auth/getAccessToken?';
-    const params = new URLSearchParams();
-    if (clientId) {
-      params.append('clientId', clientId);
-    }
-    if (scopes.length) {
-      params.append('scopes', JSON.stringify(scopes));
-    }
-    if (targetResource) {
-      params.append('targetResource', targetResource);
-    }
-    url += params.toString();
+      let url = '/api/auth/getAccessToken?';
+      const params = new URLSearchParams();
+      if (targetResource) {
+        params.append('targetResource', targetResource);
+      }
+      url += params.toString();
+
+      const result = await fetch(url, { method: 'GET', headers: { 'X-CSRF-Token': __csrf__ } });
+      const { accessToken = '' } = await result.json();
+      return accessToken;
+    } else if (authConfig.clientId && authConfig.redirectUrl && authConfig.tenantId) {
+      // get access token from cache
+      const key = authConfig.clientId + JSON.stringify(scopes);
+      let token = getTokenFromCache(key);
+      if (token && !isTokenExpired(token)) {
+        return token;
+      }
+
+      // get id token
+      if (!idToken) {
+        // pop up window if token not exist
+        const popup = createPopupWindow(
+          getIdTokenUrl({ clientId: authConfig.clientId, redirectUrl: authConfig.redirectUrl })
+        );
+        if (popup) {
+          idToken = await monitorWindowForQueryParam(popup, 'id_token', authConfig.redirectUrl);
+          storage.set('idToken', idToken || '');
+          console.log('idtoken', idToken);
```

Comment:
The token is logged to console, is it for local debugging?

Reply:
removed